File: src/tuner/services/math_expression_evaluator.py
"""Evaluates TunerStudio/Speeduino formula output channel expressions.

Formula output channels come from lines of the form
``name = { expression } [, "units"] [, digits]`` inside the ``[OutputChannels]``
section of the INI. Parsing is handled upstream by
``IniParser._parse_output_channels`` and persisted on
``EcuDefinition.formula_output_channels``. This module is the **evaluator
half** (G4 Phase 2) — it takes a formula string plus a live channel snapshot
and returns a numeric value.

Grammar — superset of ``VisibilityExpressionService`` that adds arithmetic,
unary minus, C-style bit-shifts, and the ``? :`` ternary operator so the full
vocabulary used by the production Speeduino INI (~65 formulas) evaluates
correctly::

    expr     ::= ternary
    ternary  ::= or_expr ( '?' expr ':' expr )?
    or_expr  ::= and_expr ( '||' and_expr )*
    and_expr ::= not_expr ( '&&' not_expr )*
    not_expr ::= '!' not_expr | cmp_expr
    cmp_expr ::= shift ( cmp_op shift )?
    cmp_op   ::= '==' | '!=' | '>=' | '<=' | '>' | '<'
    shift    ::= add_expr ( ('<<' | '>>') add_expr )*
    add_expr ::= mul_expr ( ('+' | '-') mul_expr )*
    mul_expr ::= unary ( ('*' | '/' | '%') unary )*
    unary    ::= ('-' | '+') unary | atom
    atom     ::= '(' expr ')' | IDENT '(' arglist ')' | NUMBER | IDENT
    arglist  ::= (expr (',' expr)*)?

Semantics mirror ``VisibilityExpressionService``:

- Booleans are floats: ``0.0`` is false, anything else is true.
- Unknown identifiers default to ``0.0``.
- Division by zero returns ``0.0`` (fail-safe, never ``inf``/``nan``).
- Bit-shift operands are coerced to ``int`` before the shift, result is float.
- Any parse error returns ``0.0`` (fail-safe — a broken formula must not
  leak exceptions into the runtime poll tick).

The only supported function is ``arrayValue(arrayName, indexExpr)`` with the
same semantics as the visibility service (optional ``array.`` prefix, integer
truncation of the index, out-of-range returns ``0.0``).
"""
from __future__ import annotations

import re

from tuner.domain.ecu_definition import EcuDefinition, FormulaOutputChannel
from tuner.domain.output_channels import OutputChannelSnapshot, OutputChannelValue


_TOKEN_RE = re.compile(
    r"(?P<NUMBER>\d+(?:\.\d+)?)"
    r"|(?P<OP><<|>>|==|!=|>=|<=|&&|\|\||[+\-*/%><!(),?:])"
    r"|(?P<IDENT>[A-Za-z_]\w*(?:\.[A-Za-z_]\w*)*)"
    r"|(?P<WS>\s+)"
)


def _tokenize(text: str) -> list[str]:
    return [m.group() for m in _TOKEN_RE.finditer(text) if m.lastgroup != "WS"]


class _Parser:
    __slots__ = ("_tokens", "_pos", "_values", "_arrays")

    def __init__(
        self,
        tokens: list[str],
        values: dict[str, float],
        arrays: dict[str, list[float]] | None,
    ) -> None:
        self._tokens = tokens
        self._pos = 0
        self._values = values
        self._arrays = arrays

    def _peek(self) -> str | None:
        return self._tokens[self._pos] if self._pos < len(self._tokens) else None

    def _consume(self) -> str:
        token = self._tokens[self._pos]
        self._pos += 1
        return token

    # -- expression levels -------------------------------------------------

    def parse_expr(self) -> float:
        return self._parse_ternary()

    def _parse_ternary(self) -> float:
        cond = self._parse_or()
        if self._peek() == "?":
            self._consume()
            then_val = self._parse_ternary()
            if self._peek() == ":":
                self._consume()
            else_val = self._parse_ternary()
            return then_val if cond != 0.0 else else_val
        return cond

    def _parse_or(self) -> float:
        left = self._parse_and()
        while self._peek() == "||":
            self._consume()
            right = self._parse_and()
            left = 1.0 if (left or right) else 0.0
        return left

    def _parse_and(self) -> float:
        left = self._parse_not()
        while self._peek() == "&&":
            self._consume()
            right = self._parse_not()
            left = 1.0 if (left and right) else 0.0
        return left

    def _parse_not(self) -> float:
        if self._peek() == "!":
            self._consume()
            val = self._parse_not()
            return 0.0 if val else 1.0
        return self._parse_cmp()

    def _parse_cmp(self) -> float:
        left = self._parse_shift()
        op = self._peek()
        if op in ("==", "!=", ">=", "<=", ">", "<"):
            self._consume()
            right = self._parse_shift()
            if op == "==":
                return 1.0 if left == right else 0.0
            if op == "!=":
                return 1.0 if left != right else 0.0
            if op == ">":
                return 1.0 if left > right else 0.0
            if op == "<":
                return 1.0 if left < right else 0.0
            if op == ">=":
                return 1.0 if left >= right else 0.0
            if op == "<=":
                return 1.0 if left <= right else 0.0
        return left

    def _parse_shift(self) -> float:
        left = self._parse_add()
        while self._peek() in ("<<", ">>"):
            op = self._consume()
            right = self._parse_add()
            li = int(left)
            ri = int(right)
            if op == "<<":
                left = float(li << ri) if ri >= 0 else 0.0
            else:
                left = float(li >> ri) if ri >= 0 else 0.0
        return left

    def _parse_add(self) -> float:
        left = self._parse_mul()
        while self._peek() in ("+", "-"):
            op = self._consume()
            right = self._parse_mul()
            left = left + right if op == "+" else left - right
        return left

    def _parse_mul(self) -> float:
        left = self._parse_unary()
        while self._peek() in ("*", "/", "%"):
            op = self._consume()
            right = self._parse_unary()
            if op == "*":
                left = left * right
            elif op == "/":
                left = (left / right) if right != 0.0 else 0.0
            else:  # %
                left = (left % right) if right != 0.0 else 0.0
        return left

    def _parse_unary(self) -> float:
        tok = self._peek()
        if tok == "-":
            self._consume()
            return -self._parse_unary()
        if tok == "+":
            self._consume()
            return self._parse_unary()
        return self._parse_atom()

    def _parse_atom(self) -> float:
        token = self._peek()
        if token is None:
            return 0.0
        if token == "(":
            self._consume()
            val = self._parse_ternary()
            if self._peek() == ")":
                self._consume()
            return val
        self._consume()
        # Function call?
        if self._peek() == "(":
            return self._call_function(token)
        try:
            return float(token)
        except ValueError:
            return float(self._values.get(token, 0.0))

    def _call_function(self, name: str) -> float:
        self._consume()  # consume '('
        if name == "arrayValue":
            return self._eval_array_value()
        # Unknown function — skip to matching ')' and return 0.0.
        depth = 1
        while self._peek() is not None and depth > 0:
            tok = self._consume()
            if tok == "(":
                depth += 1
            elif tok == ")":
                depth -= 1
        return 0.0

    def _eval_array_value(self) -> float:
        """Evaluate arrayValue(arrayName, indexExpr) — already consumed '('."""
        array_name_token = self._peek()
        if array_name_token not in (None, ")", ","):
            self._consume()
        else:
            array_name_token = ""
        if self._peek() == ",":
            self._consume()
        index_val = self._parse_ternary()
        if self._peek() == ")":
            self._consume()

        name = (array_name_token or "").strip()
        if name.startswith("array."):
            name = name[len("array."):]
        if not name or self._arrays is None:
            return 0.0
        arr = self._arrays.get(name)
        if arr is None:
            return 0.0
        index = int(index_val)
        if 0 <= index < len(arr):
            return float(arr[index])
        return 0.0


class MathExpressionEvaluator:
    """Evaluates TunerStudio-style formula output channel expressions.

    This is the evaluator half of G4 (virtual / formula output channels). The
    parser half lives in ``IniParser._parse_output_channels`` and persists
    results on ``EcuDefinition.formula_output_channels``.
    """

    def evaluate(
        self,
        expression: str | None,
        values: dict[str, float],
        arrays: dict[str, list[float]] | None = None,
    ) -> float:
        """Return the numeric value of *expression* against *values*.

        Parameters
        ----------
        expression:
            Raw expression string. The surrounding ``{ ... }`` braces (as
            persisted by the INI parser for visibility expressions) are
            accepted for symmetry with ``VisibilityExpressionService`` even
            though formula channels store the inner form on
            ``FormulaOutputChannel.formula_expression``.
        values:
            Current channel values keyed by channel name (live output
            channel snapshot plus any already-computed formula channels).
        arrays:
            Optional ``arrayValue(name, index)`` lookup table.

        Returns
        -------
        float
            The computed value, or ``0.0`` if *expression* is empty or a
            parse error occurs.
        """
        if not expression:
            return 0.0
        expr = expression.strip()
        if expr.startswith("{") and expr.endswith("}"):
            expr = expr[1:-1].strip()
        if not expr:
            return 0.0
        try:
            tokens = _tokenize(expr)
            parser = _Parser(tokens, values, arrays)
            return parser.parse_expr()
        except Exception:
            return 0.0

    def enrich_snapshot(
        self,
        snapshot: OutputChannelSnapshot,
        definition: EcuDefinition | None,
    ) -> OutputChannelSnapshot:
        """Return *snapshot* with formula output channels appended.

        The input snapshot is **not** mutated. Existing hardware channels
        are preserved in order; freshly-computed formula channels are
        appended after them, using ``FormulaOutputChannel.units`` as each
        computed entry's units.

        If *definition* is ``None`` or has no formula channels, the input
        snapshot is returned unchanged (same object — no copy). Runtime
        paths can call this unconditionally on every poll tick without
        paying an allocation when there are no formulas.

        The definition's ``output_channel_arrays`` map is passed through
        as the ``arrays`` argument to ``compute_all`` so formulas that
        reference ``arrayValue(array.name, index)`` can resolve cleanly.
        """
        if definition is None or not definition.formula_output_channels:
            return snapshot
        computed = self.compute_all(
            definition.formula_output_channels,
            snapshot.as_dict(),
            definition.output_channel_arrays or None,
        )
        units_by_name = {
            f.name: f.units for f in definition.formula_output_channels
        }
        appended = list(snapshot.values)
        for f in definition.formula_output_channels:
            appended.append(
                OutputChannelValue(
                    name=f.name,
                    value=computed[f.name],
                    units=units_by_name.get(f.name),
                )
            )
        return OutputChannelSnapshot(
            timestamp=snapshot.timestamp, values=appended
        )

    def compute_all(
        self,
        formulas: list[FormulaOutputChannel],
        values: dict[str, float],
        arrays: dict[str, list[float]] | None = None,
    ) -> dict[str, float]:
        """Compute every formula channel against *values* in declaration order.

        Each computed value is folded back into the working snapshot so
        later formulas can reference earlier ones (the production INI does
        this: ``cycleTime = revolutionTime * strokeMultipler`` depends on
        ``revolutionTime`` and ``strokeMultipler`` being computed first).

        The input *values* dict is **not** mutated. The returned dict
        contains only the freshly-computed formula channels.
        """
        working: dict[str, float] = dict(values)
        computed: dict[str, float] = {}
        for f in formulas:
            result = self.evaluate(f.formula_expression, working, arrays)
            computed[f.name] = result
            working[f.name] = result
        return computed

File: cpp/src/math_expression_evaluator.cpp
// SPDX-License-Identifier: MIT
//
// tuner_core::math_expression_evaluator implementation. Direct port of
// ``tuner.services.math_expression_evaluator`` — pure logic, stdlib only.

#include "tuner_core/math_expression_evaluator.hpp"

#include <cctype>
#include <cmath>
#include <cstdint>
#include <stdexcept>
#include <string>
#include <vector>

namespace tuner_core::math_expression_evaluator {

namespace {

// ---------------------------------------------------------------------------
// Tokenizer
// ---------------------------------------------------------------------------

// Token kinds mirror the Python regex:
//   NUMBER  → digits, optional '.' + digits
//   OP      → <<, >>, ==, !=, >=, <=, &&, ||, +, -, *, /, %, >, <, !,
//             (, ), ',', '?', ':'
//   IDENT   → [A-Za-z_]\w*(?:\.[A-Za-z_]\w*)*
// Whitespace is dropped. Unrecognized characters are silently skipped.
std::vector<std::string> tokenize(std::string_view text) {
    std::vector<std::string> tokens;
    const std::size_t n = text.size();
    std::size_t i = 0;
    while (i < n) {
        const char c = text[i];
        if (std::isspace(static_cast<unsigned char>(c))) {
            ++i;
            continue;
        }
        // Number
        if (std::isdigit(static_cast<unsigned char>(c))) {
            std::size_t j = i;
            while (j < n && std::isdigit(static_cast<unsigned char>(text[j]))) ++j;
            if (j < n && text[j] == '.') {
                ++j;
                while (j < n && std::isdigit(static_cast<unsigned char>(text[j]))) ++j;
            }
            tokens.emplace_back(text.substr(i, j - i));
            i = j;
            continue;
        }
        // Two-char operators
        if (i + 1 < n) {
            const char d = text[i + 1];
            if ((c == '<' && d == '<') ||
                (c == '>' && d == '>') ||
                (c == '=' && d == '=') ||
                (c == '!' && d == '=') ||
                (c == '>' && d == '=') ||
                (c == '<' && d == '=') ||
                (c == '&' && d == '&') ||
                (c == '|' && d == '|')) {
                tokens.emplace_back(text.substr(i, 2));
                i += 2;
                continue;
            }
        }
        // Single-char operators
        if (c == '+' || c == '-' || c == '*' || c == '/' || c == '%' ||
            c == '>' || c == '<' || c == '!' ||
            c == '(' || c == ')' || c == ',' ||
            c == '?' || c == ':') {
            tokens.emplace_back(1, c);
            ++i;
            continue;
        }
        // Identifier (with optional `.IDENT` segments)
        if (std::isalpha(static_cast<unsigned char>(c)) || c == '_') {
            std::size_t j = i;
            auto consume_word = [&]() {
                while (j < n) {
                    char k = text[j];
                    if (std::isalnum(static_cast<unsigned char>(k)) || k == '_') {
                        ++j;
                    } else {
                        break;
                    }
                }
            };
            consume_word();
            while (j + 1 < n && text[j] == '.' &&
                   (std::isalpha(static_cast<unsigned char>(text[j + 1])) ||
                    text[j + 1] == '_')) {
                ++j;
                consume_word();
            }
            tokens.emplace_back(text.substr(i, j - i));
            i = j;
            continue;
        }
        // Unrecognized — skip silently (mirrors Python finditer).
        ++i;
    }
    return tokens;
}

// ---------------------------------------------------------------------------
// Parser
// ---------------------------------------------------------------------------

class Parser {
public:
    Parser(const std::vector<std::string>& tokens,
           const ValueMap& values,
           const ArrayMap* arrays)
        : tokens_(tokens), pos_(0), values_(values), arrays_(arrays) {}

    double parse_expr() { return parse_ternary(); }

private:
    const std::vector<std::string>& tokens_;
    std::size_t pos_;
    const ValueMap& values_;
    const ArrayMap* arrays_;

    const std::string* peek() const {
        return pos_ < tokens_.size() ? &tokens_[pos_] : nullptr;
    }
    const std::string& consume() { return tokens_[pos_++]; }

    bool peek_is(const char* s) const {
        const auto* p = peek();
        return p != nullptr && *p == s;
    }

    double parse_ternary() {
        double cond = parse_or();
        if (peek_is("?")) {
            consume();
            double then_val = parse_ternary();
            if (peek_is(":")) consume();
            double else_val = parse_ternary();
            return cond != 0.0 ? then_val : else_val;
        }
        return cond;
    }

    double parse_or() {
        double left = parse_and();
        while (peek_is("||")) {
            consume();
            double right = parse_and();
            left = (left != 0.0 || right != 0.0) ? 1.0 : 0.0;
        }
        return left;
    }

    double parse_and() {
        double left = parse_not();
        while (peek_is("&&")) {
            consume();
            double right = parse_not();
            left = (left != 0.0 && right != 0.0) ? 1.0 : 0.0;
        }
        return left;
    }

    double parse_not() {
        if (peek_is("!")) {
            consume();
            double v = parse_not();
            return v != 0.0 ? 0.0 : 1.0;
        }
        return parse_cmp();
    }

    double parse_cmp() {
        double left = parse_shift();
        const auto* op = peek();
        if (op != nullptr) {
            const std::string& s = *op;
            if (s == "==" || s == "!=" || s == ">=" || s == "<=" ||
                s == ">"  || s == "<") {
                std::string opcopy = s;
                consume();
                double right = parse_shift();
                if (opcopy == "==") return left == right ? 1.0 : 0.0;
                if (opcopy == "!=") return left != right ? 1.0 : 0.0;
                if (opcopy == ">")  return left >  right ? 1.0 : 0.0;
                if (opcopy == "<")  return left <  right ? 1.0 : 0.0;
                if (opcopy == ">=") return left >= right ? 1.0 : 0.0;
                if (opcopy == "<=") return left <= right ? 1.0 : 0.0;
            }
        }
        return left;
    }

    double parse_shift() {
        double left = parse_add();
        while (peek_is("<<") || peek_is(">>")) {
            std::string op = consume();
            double right = parse_add();
            auto li = static_cast<std::int64_t>(left);
            auto ri = static_cast<std::int64_t>(right);
            if (ri < 0) {
                left = 0.0;
            } else if (op == "<<") {
                left = static_cast<double>(li << ri);
            } else {
                left = static_cast<double>(li >> ri);
            }
        }
        return left;
    }

    double parse_add() {
        double left = parse_mul();
        while (peek_is("+") || peek_is("-")) {
            std::string op = consume();
            double right = parse_mul();
            left = (op == "+") ? left + right : left - right;
        }
        return left;
    }

    double parse_mul() {
        double left = parse_unary();
        while (peek_is("*") || peek_is("/") || peek_is("%")) {
            std::string op = consume();
            double right = parse_unary();
            if (op == "*") {
                left = left * right;
            } else if (op == "/") {
                left = (right != 0.0) ? left / right : 0.0;
            } else {
                left = (right != 0.0) ? std::fmod(left, right) : 0.0;
            }
        }
        return left;
    }

    double parse_unary() {
        if (peek_is("-")) {
            consume();
            return -parse_unary();
        }
        if (peek_is("+")) {
            consume();
            return parse_unary();
        }
        return parse_atom();
    }

    double parse_atom() {
        const auto* tok = peek();
        if (tok == nullptr) return 0.0;
        if (*tok == "(") {
            consume();
            double v = parse_ternary();
            if (peek_is(")")) consume();
            return v;
        }
        std::string token = consume();
        if (peek_is("(")) {
            return call_function(token);
        }
        try {
            std::size_t consumed = 0;
            double v = std::stod(token, &consumed);
            if (consumed == token.size()) return v;
        } catch (...) {
            // fall through
        }
        auto it = values_.find(token);
        return it != values_.end() ? it->second : 0.0;
    }

    double call_function(const std::string& name) {
        consume();  // consume '('
        if (name == "arrayValue") return eval_array_value();
        // Unknown function — skip to matching ')' and return 0.0.
        int depth = 1;
        while (peek() != nullptr && depth > 0) {
            const std::string& tok = consume();
            if (tok == "(") ++depth;
            else if (tok == ")") --depth;
        }
        return 0.0;
    }

    double eval_array_value() {
        std::string array_name;
        const auto* tok = peek();
        if (tok != nullptr && *tok != ")" && *tok != ",") {
            array_name = consume();
        }
        if (peek_is(",")) consume();
        double index_val = parse_ternary();
        if (peek_is(")")) consume();

        std::string name = array_name;
        const std::string prefix = "array.";
        if (name.rfind(prefix, 0) == 0) name = name.substr(prefix.size());
        if (name.empty() || arrays_ == nullptr) return 0.0;
        auto it = arrays_->find(name);
        if (it == arrays_->end()) return 0.0;
        const auto& arr = it->second;
        auto index = static_cast<std::int64_t>(index_val);
        if (index >= 0 &&
            static_cast<std::size_t>(index) < arr.size()) {
            return arr[static_cast<std::size_t>(index)];
        }
        return 0.0;
    }
};

std::string_view strip(std::string_view s) {
    while (!s.empty() && std::isspace(static_cast<unsigned char>(s.front()))) {
        s.remove_prefix(1);
    }
    while (!s.empty() && std::isspace(static_cast<unsigned char>(s.back()))) {
        s.remove_suffix(1);
    }
    return s;
}

}  // namespace

double evaluate(
    std::string_view expression,
    const ValueMap& values,
    const ArrayMap* arrays) {
    if (expression.empty()) return 0.0;
    auto trimmed = strip(expression);
    if (trimmed.size() >= 2 && trimmed.front() == '{' && trimmed.back() == '}') {
        trimmed = strip(trimmed.substr(1, trimmed.size() - 2));
    }
    if (trimmed.empty()) return 0.0;

    try {
        auto tokens = tokenize(trimmed);
        Parser parser(tokens, values, arrays);
        return parser.parse_expr();
    } catch (...) {
        return 0.0;
    }
}

ValueMap compute_all(
    const std::vector<IniFormulaOutputChannel>& formulas,
    const ValueMap& values,
    const ArrayMap* arrays) {
    ValueMap working = values;
    ValueMap computed;
    for (const auto& f : formulas) {
        double result = evaluate(f.formula_expression, working, arrays);
        computed[f.name] = result;
        working[f.name] = result;
    }
    return computed;
}

void enrich(
    ValueMap& working,
    const std::vector<IniFormulaOutputChannel>& formulas,
    const ArrayMap* arrays) {
    for (const auto& f : formulas) {
        if (working.find(f.name) != working.end()) {
            // Hardware channel with the same name wins — don't clobber.
            continue;
        }
        working[f.name] = evaluate(f.formula_expression, working, arrays);
    }
}

}  // namespace tuner_core::math_expression_evaluator
